### PR TITLE
Fix block gap in group blocks in classic themes

### LINKED ISSUE #47386: Block gap does not work for groups in classic themes
### Description

In https://core.trac.wordpress.org/changeset/55067 in preparation for WordPress 6.2, appearance-tools theme support was made available
With the theme support, _but without theme.json_, the block gap setting does not work with the group block, because the CSS is only applied to the inner container div, the `wp-block-group__inner-container`, not the inner blocks.

Update February 3: This affects **both** classic themes and block themes with the above condition.

(The block spacing works with the row and stack variation as these do not have the inner container)

I do not know if there is an easy CSS fix for this.
I am thinking that the easiest solution maybe to revert the change in 55067 to include the appearance tools minus block gap,
but then users will not be able to take advantage of this feature for other blocks.

### Block Gap Testing Instructions
Enable the appearance-tools theme support in your test theme.
In the block editor, add a group block with three paragraphs.
In the Dimensions panel in the block settings sidebar, set block spacing to seven.
Take note if the space between the paragraphs changes.
Save and view the space on the front.

### Testing results:
In emptytheme (block theme), the block gap correctly increases the spacing.
CSS applied to paragraph one:
```
.wp-block-group.wp-container-1 > * {
	margin-block-start: 0;
	margin-block-end: 0;
}
```
CSS applied to paragraph two:
```
.wp-block-group.wp-container-1.wp-block-group.wp-container-1 > * + * {
	margin-block-start: var(--wp--preset--spacing--80);
	margin-block-end: 0;
}
```
CSS applied to paragraph three:
```
.wp-block-group.wp-container-1.wp-block-group.wp-container-1 > * + * {
	margin-block-start: var(--wp--preset--spacing--80);
	margin-block-end: 0;
}
```

Twenty Twenty: The block gap does not work on the default group.
CSS applied to the inner container:
```
wp-block-group.wp-container-2 > * {
	margin-block-start: 0;
	margin-block-end: 0;
}
```
Paragraph one, Paragraph two, Paragraph three:
No wp-container-2 CSS is applied.

Twenty Twenty-One: The block gap does not work on the default group.
CSS applied to the inner container:
```
wp-block-group.wp-container-2 > * {
	margin-block-start: 0;
	margin-block-end: 0;
}
```
Paragraph one, Paragraph two, Paragraph three:
No wp-container-2 CSS is applied.

Twenty Sixteen ( tested for comparison )
CSS applied to the inner container:
```
wp-block-group.wp-container-1 > * {
	margin-block-start: 0;
	margin-block-end: 0;
}
```
Paragraph one, Paragraph two, Paragraph three:
No wp-container-1 CSS is applied.

Hello Elementor ( tested for comparison ): The block gap does not work on the default group.
CSS applied to the inner container:
```
wp-block-group.wp-container-1 > * {
	margin-block-start: 0;
	margin-block-end: 0;
}
```
Paragraph one, Paragraph two, Paragraph three:
No wp-container-1 CSS is applied.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

WordPress 6.2 Alpha
Tested with and without Gutenberg current trunk.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by carolinan**
## What?

In preparation for WordPress 6.2, appearance-tools theme support was made available.

With the theme support, _but without theme.json_, the block gap setting does not work with the group block, because the CSS is only applied to the inner container div, the `wp-block-group__inner-container`, not the inner blocks.
Update: This affects **both** classic themes and block themes with the above condition.

This PR adds the `.wp-block-group__inner-container` as an additional partial selector for block gap.

Closes https://github.com/WordPress/gutenberg/issues/47386

## Why?

To allow classic themes to take advantage of new features.

## Testing Instructions
With WordPress 6.2 alpha:

1. Install and activate Twenty Twenty.
2. Add `add_theme_support( 'appearance-tools');` to the themes setup function in functions.php.
3. In the block editor, add one each of group, stack variation and row variation. (The latter two to confirm that nothing broke)
4. Set the block spacing setting to something visible, like 7.
5. Check if the block spacing is working. The editor and front should match.

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
Thanks for opening this PR @carolinan!

I'm wondering if we could support classic themes that opt in to appearance tools in a slightly different way. What if instead of adding additional layout rules, with the presence of the appearance tools setting, we instead skipped the [`wp_restore_group_inner_container`](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/block-supports/layout.php#L460) behaviour (i.e. add a check to that function and return early)?

The support for the legacy group inner container markup was so that classic themes aren't adversely affected by new features in core/Gutenberg — however, if themes are opting in to use the appearance tools, I think it'd be good if they were also using the new markup, as it will help to keep the layout support a little more streamlined.

What do you think?

**Comment by carolinan** [comment]
I don't know how that would be possible in an already existing theme, where users already have been able to target this element with custom CSS. For example the bundled themes.

**Comment by andrewserong** [comment]
Ah, gotcha. In that case, though, if the themes are wishing to opt-in to new features, shouldn't they be decoupling from the legacy group block markup before adopting appearance tools? My understanding was that the goal for the back-compat code for legacy group blocks was to give themes more time before adopting new features. I think it'd be preferable to avoid that back-compat code growing to include additional rules being output by the layout support, as they're rules that would need to be maintained and would be redundant on block themes.

Apologies if I'm missing something, I haven't seen the discussions surrounding the appearance tools feature for classic themes, but my preference would be to try to avoid adding additional complexity to the layout support if it isn't needed, since the rules are already quite complex.

**Comment by carolinan** [comment]
I don't understand what "decoupling from the legacy markup" means and if that is something I should be able to do with a PR to the bundled themes.

**Comment by andrewserong** [comment]
> I don't understand what "decoupling from the legacy markup" means and if that is something I should be able to do with a PR to the bundled themes.

Apologies, I'm not really too sure what the path forward would be here, as I'm still trying to understand the objective behind opting in to blockGap for Classic themes, as it seems like the wrong way around to me. My opinion is that for features that are designed for sites with `theme.json` files, such as block gap, it would be better for those themes wishing to adopt the features to iteratively adopt support for the blocks-theme way of doing things rather than add in extra code to the layout support for classic themes.

My main concerns are about the maintainability of the layout support, and the addition of code to target classic themes. Perhaps it would be better for the appearance tools feature in classic themes to skip blockGap?

**Comment by carolinan** [comment]
My main concern is that if blocks in the bundled classic themes are broken, that will discourage users from switching to using blocks. And broken blocks are what they are experiencing and reporting now. 

The main problem with supporting blocks in the classic themes is not blockgap. 
I think it would be a shame to not enable users to take advantage of it. But what is actually breaking right now is the other appearance tools.

**Comment by carolinan** [comment]
I would still like to understand what negative effects adding this selector would have. It is a small code change for a rather big win for users of classic themes.
It is also only affecting one block, and only one variation for that block.

**Comment by andrewserong** [comment]
> I would still like to understand what negative effects adding this selector would have.

That's a very reasonable question. I'll try to share where I'm coming from. There are a few things to evaluate when deciding whether or not to introduce a change like this beyond the size of the code itself. While the size of the change is fairly minimal on the surface, some of the concerns that come to mind are:

* The ability to change something once it's introduced
* Whether the change is consistent with the intended scope of a feature and its future support
* What the feature communicates about the scope of a feature
* Whether or not folks will test for or maintain the feature once it's introduced

In this case, while the code change is small, it deviates from the intended use cases behind the feature it's a part of (layout) which was designed with block themes in mind, and it expands the scope of the backwards compatibility area of layout, which is an area that folks are attempting to reduce rather than increase.

So part of my concern is that this PR is less of a bug fix, and more an enhancement request, as it increases the intended scope of the block support. The way that's been recommended for themes to start using layout and block spacing features up until now has been to adopt `theme.json` — if that strategy has changed, then I think it would be better to look at the strategy behind design tools more holistically rather than adding in additional rules for classic themes, as once they're added, it likely won't possible to remove them.

This is all just my opinion, though, from working on the layout block support for a while, and running into lots of edge cases with it as it's a very complex feature. Small additions might appear to be simple in isolation, however I believe we need to be careful whenever we increase the scope to ensure that the direction is consistent with future intended usage.

Very happy to hear others' feedback if there are more opinions on all this!

**Comment by tellthemachines** [comment]
It's unclear what the best solution might be here because this is a feature that tries to meld two incompatible paradigms: classic themes and block themes. It may work well for design tools that apply to a single block, such as border, padding and line-height, but for layout-related controls such as block gap (and I'd argue even margin) it's not ideal, because classic themes already define such styles themselves. 

I understand that it's useful to have Group, Row and Stack blocks in classic themes, but that tension between paradigms keeps popping up in the form of bugs and unexpected behaviour. For instance, Row block doesn't work at all in TT1 due to an overly-specific theme style.

But these thoughts can hardly help to solve the problem at hand 😅 

I agree with @andrewserong that adding extra styles to the layout logic is not great, mainly due to the back-compat nature of the Group inner wrapper, and the transitional nature of the appearance tools support feature meaning this is a bit of an edge case.

Ideally, we would be able to scrap the group inner wrapper. But if we can't, how about adding the layout classes on the inner wrapper instead? We could add it somewhere in [here](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L590), so the logic to support the edge case scenario will all be in one place.

**Comment by carolinan** [comment]
The feature, the theme support, was added in August: https://github.com/WordPress/gutenberg/pull/43337

**Comment by andrewserong** [comment]
> But if we can't, how about adding the layout classes on the inner wrapper instead? We could add it somewhere in [here](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L590), so the logic to support the edge case scenario will all be in one place.

That sounds like a better compromise to me, in that the layout rules would be unchanged, it's just where they're being attached in Classic themes, so could be a better way to contain it.

**Comment by carolinan** [comment]
I think it is only this class that needs to be moved? `is-layout-constrained`

**Comment by andrewserong** [comment]
> I think it is only this class that needs to be moved? is-layout-constrained

If you're injecting layout classnames to group, you might also need `is-layout-flow` since group can either be full width (flow) or constrained (also, would you need `is-layout-flex` if Row and Stack need to be considered?). You could probably manually look for whether the `layout.type` block attribute is set and output the hard-coded classname (either `is-layout-flow` or `is-layout-constrained`) in `gutenberg_restore_group_inner_container` (around where the inner container classname is output).

If you need to do a look-up from the layout definitions, though, another approach might be to borrow some of code from `gutenberg_render_layout_support_flag` to generate the classname from within `gutenberg_restore_group_inner_container`. It could be something like the following (I haven't tested this, this is just a quick copy / paste / edit 🙂):

```php
	$global_settings = gutenberg_get_global_settings();
	$global_layout_settings = _wp_array_get( $global_settings, array( 'layout' ), null );
	$layout_definitions = _wp_array_get( $global_layout_settings, array( 'definitions' ), array() );
	$layout_type = _wp_array_get( $block, array( 'attrs', 'layout', 'type' ), null );
	if ( $layout_type && ! empty( $layout_definitions ) ) {
		$layout_classname = _wp_array_get( $layout_definitions, array( $layout_type, 'className' ), '' );
	} else {
		$layout_classname = _wp_array_get( $layout_definitions, array( 'default', 'className' ), '' );
	}
```

**Comment by carolinan** [comment]
The appearance tools enables margin, padding, block gap but not layout.
I think that means the flow class is never used?

Row and stack were implemented without the inner div, so they already work well with block gap.

**Comment by carolinan** [comment]
Only the default 24px spacing is working, and this should have been obvious to me:` wp-container-` class is on the outer div, and the gap value from the block setting is never output.

**Comment by carolinan** [comment]
...Or should I be using this as a reference? https://github.com/WordPress/gutenberg/pull/44600/files

**Comment by andrewserong** [comment]
> and this should have been obvious to me: wp-container- class is on the outer div, and the gap value from the block setting is never output.

Ah, of course!

> ...Or should I be using this as a reference?

The HTML tag processor would be a neat way to do this actually, then you wouldn't need to construct the layout classnames manually. It should be possible to use `get_attribute` to grab the classnames from the outer wrapper, and then insert them to the inner wrapper. One potential caveat is that the tag processor is still under review for landing in core. For this Gutenberg PR, I think it's still a good time to use the tag processor, but I don't know if it's 100% confirmed that the tag processor will land in time for 6.2 (though the hope is that it will).

**Comment by carolinan** [comment]
I think it is better if we make a decision early to [revert the appearance-tools theme support](https://core.trac.wordpress.org/ticket/57460#comment:15) in core than to stress the solution.
The theme support is still in Gutenberg. So classic themes that use Gutenberg can still opt-in. But I don't think there is a big awareness of this, because there has been no error reports from users/developers. These problems were only found when testing for them specifically...

**Comment by ndiego** [comment]
@carolinan now that appearance tools are getting reverted via https://core.trac.wordpress.org/ticket/57649. Are you ok with me punting this to 6.3 on the [project board](https://github.com/orgs/WordPress/projects/63/views/1)?

**Comment by carolinan** [comment]
> @carolinan now that appearance tools are getting reverted via https://core.trac.wordpress.org/ticket/57649. Are you ok with me punting this to 6.3 on the [project board](https://github.com/orgs/WordPress/projects/63/views/1)?

Yes, please, we need to discuss it further.

**Comment by carolinan** [comment]
So first step would be to decide if block gap should be supported in themes without theme.json.
From the conversations above, I believe the answer is no; it was never intended for these themes.

**Comment by ndiego** [comment]
Hey @carolinan, this PR is on the [6.3 project board](https://github.com/orgs/WordPress/projects/103/views/1) since it was punted from 6.2, but it doesn't look like any progress has been made. It is ok if I remove it from the 6.3 board as well?

**Comment by carolinan** [comment]
@ndiego The problem is that the underlying issue still needs to be solved. I still need the project's decision-makers to look into the issue and decide whether classic themes can regain features like the link color controls. By removing it from the board the issue is only going to be more invisible.

**Comment by ndiego** [comment]
> The problem is that the underlying issue still needs to be solved. I still need the project's decision-makers to look into the issue and decide whether classic themes can regain features like the link color controls. By removing it from the board the issue is only going to be more invisible.

cc @tellthemachines @ramonjd

**Comment by carolinan** [comment]
TLDR
- The link color and border controls have their own theme support added in 6.3 and 16.2.
- ` add_theme_support( 'appearance-tools');` which is already supported in Gutenberg, still enables block gap which is broken for the group block (and only the group block, it works elsewhere).

The remaining question is if that should be fixed, or if the block gap can be excluded from the theme support.

**Comment by tellthemachines** [comment]
This is not a straightforward bug fix, and it's not a regression from the previous WP release either, so it should be punted to the next release.

**Comment by ramonjd** [comment]
I've been testing this and noticed couple of things:

With the changes in this PR, Row and Stack do display block gap correctly in classic themes:

Screenshot 2023-07-04 at 11 00 13 am

I might be doing something wrong, but the regular Group block however (constrained) isn't working for me. It looks like `wp-block-group__inner-container` needs to `wp-container-5`, or at least the styles:

```css
.wp-container-5.wp-container-5.wp-container-5.wp-container-5 > * + * {
    margin-block-start: 100px;
    margin-block-end: 0;
}
````

To be honest I'm not confident enough to say either way. @tellthemachines and @andrewserong have good judgement in this area, which tells me it can probably be postponed until 6.4. That is, work on a fix in Gutenberg and stabilize it until then.

**Comment by ramonjd** [comment]
> This is not a straightforward bug fix, and it's not a regression from the previous WP release either, so it should be punted to the next release.

Oh, just saw this was posted after I hit "Comment".

Thanks @tellthemachines

**Comment by carolinan** [comment]
So is the consensus now that the block gap should be enabled for themes without theme.json?

**Comment by carolinan** [comment]
This is not resolved, it is still unclear if the feature should be enabled for classic themes or not. Removing it from the 6.4 board.

File: lib/block-supports/layout.php [hunk 1]
```diff
@@ -836,14 +837,22 @@ function gutenberg_restore_group_inner_container( $block_content, $block ) {
...
+
 	$replace_regex   = sprintf(
 		'/(^\s*<%1$s\b[^>]*wp-block-group[^>]*>)(.*)(<\/%1$s>\s*$)/ms',
 		preg_quote( $tag_name, '/' )
 	);
 	$updated_content = preg_replace_callback(
 		$replace_regex,
-		static function ( $matches ) {
-			return $matches[1] . '<div class="wp-block-group__inner-container">' . $matches[2] . '</div>' . $matches[3];
+		static function( $matches ) use ( $constrained_class ) {
```
Comment (carolinan) [inline lib/block-supports/layout.php]: ``
```suggestion
		static function ( $matches ) use ( $constrained_class ) {
```

**Comment by tellthemachines** [comment]
Now that https://github.com/WordPress/gutenberg/pull/56130 has been merged, we can close this PR. Thanks for all your work and patience with the endless back and forth on this one @carolinan!
